crates/loro-internal/src/kv_store/mem_store.rs

```diff
@@ -0,0 +1,395 @@
+//! # MemKvStore Documentation
+//!
+//! MemKvStore use SSTable as backend. The SSTable (Sorted String Table) is a persistent data structure used for storing key-value pairs in a sorted manner. This document describes the binary format of the SSTable.
+//!
+//! ## Overall Structure
+//!
+//! The SSTable consists of the following sections:
+//!
+//! ```
+//! ┌─────────────────────────────────────────────────────────────────────────────────────────────────┐
+//! │ MemKVStore                                                                                      │
+//! │┌ ─ ─ ─ ─ ─ ─ ─┌ ─ ─ ─ ─ ─ ─ ─ ─┌ ─ ─ ─ ─ ─ ─ ┬ ─ ─ ─┌ ─ ─ ─ ─ ─ ─ ─ ┬ ─ ─ ─ ─ ─ ─┌ ─ ─ ─ ─ ─ ─ ┐│
+//! │  Magic Number │ Schema Version │ Block Chunk   ...  │  Block Chunk    Block Meta │ Meta Offset  │
+//! ││     u32      │       u8       │    bytes    │      │     bytes     │   bytes    │     u32     ││
+//! │ ─ ─ ─ ─ ─ ─ ─ ┘─ ─ ─ ─ ─ ─ ─ ─ ┘─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┘─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┘─ ─ ─ ─ ─ ─ ─ │
+//! └─────────────────────────────────────────────────────────────────────────────────────────────────┘
+//! ```
+//!
+//! 1. Magic Number (4 bytes): A fixed value "LORO" to identify the file format.
+//! 2. Schema Version (1 byte): The version of the MemKVStore schema.
+//! 3. Block Chunks: A series of data blocks containing key-value pairs.
+//! 4. Block Meta: Metadata for all blocks, including block offset, first key, is_large flag, and last key if not large.
+//! 5. Meta Offset (4 bytes): The offset of the Block Meta section from the beginning of the file.
+//!
+//! ## Block Types
+//!
+//! There are two types of blocks: Normal Blocks and Large Value Blocks.
+//!
+//! ### Normal Block
+//!
+//! Normal blocks store multiple key-value pairs with compressed keys.
+//!
+//! ```
+//! ┌────────────────────────────────────────────────────────────────────────────────────────────┐
+//! │Block                                                                                   │
+//! │┌ ─ ─ ─ ─ ─ ─ ─ ┬ ─ ─ ─┌ ─ ─ ─ ─ ─ ─ ─ ┬ ─ ─ ─ ─┌ ─ ─ ─┌ ─ ─ ─ ┬ ─ ─ ─ ─┌ ─ ─ ─ ─ ─ ─ ─ │
+//! │ Key Value Chunk  ...  │Key Value Chunk  offset │ ...  │ offset  kv len │Block Checksum││
+//! ││     bytes     │      │     bytes     │  u16   │      │  u16  │  u16   │     u32       │
+//! │ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┘─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┘─ ─ ─ ┘─ ─ ─ ─ ─ ─ ─ ─ ┘─ ─ ─ ─ ─ ─ ─ ┘│
+//! └────────────────────────────────────────────────────────────────────────────────────────────┘
+//! ```
+//!
+//! Each Key Value Chunk is encoded as follows:
+//!
+//! ```
+//! ┌───────────────────────────────────────────────────────────────┐
+//! │  Key Value Chunk                                              │
+//! │┌ ─ ─ ─ ─ ─ ─ ─ ─ ┬ ─ ─ ─ ─ ─ ─ ─┌ ─ ─ ─ ─ ─┌ ─ ─ ─ ─ ┬ ─ ─ ─ ┐│
+//! │ common prefix len key suffix len│key suffix│value len  value  │
```

Comment:
It seems we don't need to encode the value len here as well.

It's a special case for the first chunk and the last chunk of the block, right? We don't need to encode the keys for them.

Reply:
I think so too, but I haven't modified it yet.